From: Samuel Mendoza-Jonas <samjonas@amazon.com>
Date: Fri, 20 Jan 2023 18:09:39 +0000
Subject: Revert "PCI/MSI: Let core code free MSI descriptors"

This reverts commit 9fb9eb4b59acc607e978288c96ac7efa917153d4.

This commit causes issues with hibernation on Xen guests (null-pointer
exception on freeze), most likey due to the Xen driver trying to work on
interrupts after they've been freed.

While the downstream Xen patches are updated to work properly, revert
this change. This is modified to still use the new accessors introduced
in the same series, and just call out to msi_free_msi_descs() which does
exactly the same work.

Signed-off-by: Samuel Mendoza-Jonas <samjonas@amazon.com>
---
 drivers/pci/msi/irqdomain.c |  1 -
 drivers/pci/msi/msi.c       | 10 ++++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/pci/msi/irqdomain.c b/drivers/pci/msi/irqdomain.c
--- a/drivers/pci/msi/irqdomain.c
+++ b/drivers/pci/msi/irqdomain.c
@@ -28,7 +28,6 @@ void pci_msi_teardown_msi_irqs(struct pci_dev *dev)
 		msi_domain_free_irqs_descs_locked(domain, &dev->dev);
 	else
 		pci_msi_legacy_teardown_msi_irqs(dev);
-	msi_free_msi_descs(&dev->dev);
 }
 
 /**
diff --git a/drivers/pci/msi/msi.c b/drivers/pci/msi/msi.c
--- a/drivers/pci/msi/msi.c
+++ b/drivers/pci/msi/msi.c
@@ -224,8 +224,18 @@ EXPORT_SYMBOL_GPL(pci_write_msi_msg);
 
 static void free_msi_irqs(struct pci_dev *dev)
 {
+	struct msi_desc *desc;
+	int i;
+
+	msi_for_each_desc(desc, &dev->dev, MSI_DESC_NOTASSOCIATED)
+		if (desc->irq)
+			for (i = 0; i < desc->nvec_used; i++)
+				BUG_ON(irq_has_action(desc->irq + i));
+
 	pci_msi_teardown_msi_irqs(dev);
 
+	msi_free_msi_descs(&dev->dev);
+
 	if (dev->msix_base) {
 		iounmap(dev->msix_base);
 		dev->msix_base = NULL;

